devito/finite_differences/derivative.py

```diff
@@ -153,35 +213,37 @@ def _xreplace(self, subs):
         substitutions until evaluation.
         """
         subs = self._subs + (subs,)  # Postponed substitutions
-        return Derivative(self.expr, *self.dims, deriv_order=self.deriv_order,
-                          fd_order=self.fd_order, side=self.side,
-                          transpose=self.transpose, subs=subs, x0=self.x0), True
+        return self._new_from_self(subs=subs), True
 
-    @property
+    @cached_property
     def dims(self):
         return self._dims
 
-    @property
+    @cached_property
+    def ndims(self):
+        return len(self._dims)
+
+    @cached_property
     def x0(self):
         return self._x0
 
-    @property
+    @cached_property
     def fd_order(self):
         return self._fd_order
 
-    @property
+    @cached_property
```

Comment:
remove all cached_properties when you're just returning a field (ie the property has 0 cost), because they will be of no help (in fact, they might increase access time and storage requirements!)